lib/internal/test_runner/runner.js

```diff
@@ -248,14 +220,79 @@ class FileTest extends Test {
       this.#buffer = [];
     }
   }
-  addToReport(ast) {
-    this.#accumulateReportItem(ast);
+  addToReport(item) {
+    this.#accumulateReportItem(item);
     if (!this.isClearToSend()) {
-      ArrayPrototypePush(this.#buffer, ast);
+      ArrayPrototypePush(this.#buffer, item);
       return;
     }
     this.#drainBuffer();
-    this.#handleReportItem(ast);
+    this.#handleReportItem(item);
+  }
+  parseMessage(readData) {
+    if (readData.length === 0) return;
+
+    ArrayPrototypePush(this.#messageBuffer, readData);
+    this.#messageBufferSize += readData.length;
+
+    // Index 0 should always be present because we just pushed data into it.
+    let messageBufferHead = this.#messageBuffer[0];
+    let headerIndex = messageBufferHead.indexOf(v8Header);
+    let nonSerialized = Buffer.alloc(0);
+
+    while (messageBufferHead && headerIndex !== kSerializedSizeHeader) {
+      const nonSerializedData = headerIndex === -1 ?
+        messageBufferHead :
+        messageBufferHead.slice(0, headerIndex - kSerializedSizeHeader);
+      nonSerialized = Buffer.concat([nonSerialized, nonSerializedData]);
+      this.#messageBufferSize -= nonSerializedData.length;
+      if (headerIndex === -1) {
+        ArrayPrototypeShift(this.#messageBuffer);
+      } else {
+        this.#messageBuffer[0] = messageBufferHead.subarray(headerIndex - kSerializedSizeHeader);
+      }
+      messageBufferHead = this.#messageBuffer[0];
+      headerIndex = messageBufferHead?.indexOf(v8Header);
+    }
+
+    if (nonSerialized.length > 0) {
+      const messages = RegExpPrototypeSymbolSplit(/\r?\n/, nonSerialized.toString('utf-8'));
+      messages.forEach((message) => {
+        this.addToReport({
+          __proto__: null,
+          type: 'test:diagnostic',
+          data: { __proto__: null, nesting: 0, file: this.name, message },
+        });
+      });
```

Comment:
`RegExpPrototypeSymbolSplit` looks up the `flags` property of the regex, which is a mutable getter. We'd need to either call `SideEffectFreeRegExpPrototypeSymbolSplit`, use `primordials.hardenRegExp`, or workaround the issue with `RegExpPrototypeExec`. The latter is probably the most performant, albeit not as readable.
```suggestion
      const needle = /\r?\n/g;
      const haystack = nonSerialized.toString('utf-8');
      let match;
      let lastIndex = 0;
      while ((match = RegExpPrototypeExec(needle, haystack)) != null) {
        this.addToReport({
          __proto__: null,
          type: 'test:diagnostic',
          data: { __proto__: null, nesting: 0, file: this.name, message: StringPrototypeSlice(heystack, lastIndex, match.index) },
        });
        lastIndex = needle.lastIndex;
      };
      if (lastIndex !== haystack.length) {
        this.addToReport({
          __proto__: null,
          type: 'test:diagnostic',
          data: { __proto__: null, nesting: 0, file: this.name, message: StringPrototypeSlice(heystack, lastIndex) },
        });
      }
```